camayoc/tests/qpc/api/v1/reports/test_reports.py

```diff
@@ -116,62 +114,63 @@ def test_products_found_deployment_report(data_provider, scan_info):
+    system_fingerprints = report_content.get("system_fingerprints")
+    found_hosts = {}
+    for host in system_fingerprints:
+        found_hosts[host.get("name")] = host
+    if len(found_hosts) != len(system_fingerprints):
+        msg = "Some discovered hosts have the same name. " "Test result might not be accurate."
+        warnings.warn(msg)
     errors_found = []
-    for entity in report:
-        all_found_products = []
-        present_products = []
-        for product in entity.get("products"):
-            name = "".join(product["name"].lower().split())
-            if product["presence"] == "present":
-                present_products.append(name)
-            if product["presence"] in ["present", "potential"]:
-                all_found_products.append(name)
-        for source_to_product_map in scan_info.get("expected_products", []):
-            src_id = list(source_to_product_map.keys())[0]
-            entity_src_ids = get_report_entity_src_ids(entity)
-            hostname = result["source_id_to_hostname"][src_id]
-            ex_products = source_to_product_map[src_id]
-            expected_product_names = [prod for prod in ex_products.keys() if prod != "distribution"]
-            if src_id in entity_src_ids:
-                # We assert that products marked as present are expected
-                # We do not assert that products marked as potential must
-                # actually be on server
-                unexpected_products = []
-                for prod_name in present_products:
-                    # Assert that products marked "present"
-                    # Are actually expected on machine
-                    if prod_name not in expected_product_names:
-                        unexpected_products.append(prod_name)
-                # after inpsecting all found products,
-                # raise assertion error for all unexpected products
-                if len(unexpected_products) > 0:
-                    errors_found.append(
-                        "Found {found_products} but only expected to find\n"
-                        "{expected_products} on {host_found_on}.\n"
-                        "All information about the entity was as follows\n"
-                        "{entity_info}".format(
-                            found_products=unexpected_products,
-                            expected_products=expected_product_names,
-                            host_found_on=hostname,
-                            entity_info=pformat(entity),
-                        )
+    for hostname, expected_data in scan_info.expected_data.items():
+        expected_products = expected_data.products
+        if not expected_products:
+            continue
+        actual_data = found_hosts.get(hostname)
+        if not actual_data:
+            errors_found.append(
+                f"Host '{hostname}' was expected for scan {scan_info.get('name')}, but not found"
+            )
```

Comment:
Since we are only supporting Python 3.11, I think we could use walrus operators here (but please check if this is correct):
```suggestion
        if not (expected_products := expected_data.products):
            continue
        if not (actual_data := found_hosts.get(hostname)):
            errors_found.append(
                f"Host '{hostname}' was expected for scan {scan_info.get('name')}, but not found"
            )
```

Reply:
Let's use it.